packages/components/src/slot-fill/provider.tsx

```diff
@@ -8,103 +8,102 @@ import { useState } from '@wordpress/element';
  */
 import SlotFillContext from './context';
 import type {
-	FillComponentProps,
+	FillInstance,
+	FillChildren,
+	BaseSlotInstance,
 	BaseSlotFillContext,
 	SlotFillProviderProps,
 	SlotKey,
-	Rerenderable,
 } from './types';
+import { observableMap } from '@wordpress/compose';
 
 function createSlotRegistry(): BaseSlotFillContext {
-	const slots: Record< SlotKey, Rerenderable > = {};
-	const fills: Record< SlotKey, FillComponentProps[] > = {};
-	let listeners: Array< () => void > = [];
-
-	function registerSlot( name: SlotKey, slot: Rerenderable ) {
-		const previousSlot = slots[ name ];
-		slots[ name ] = slot;
-		triggerListeners();
-
-		// Sometimes the fills are registered after the initial render of slot
-		// But before the registerSlot call, we need to rerender the slot.
-		forceUpdateSlot( name );
-
-		// If a new instance of a slot is being mounted while another with the
-		// same name exists, force its update _after_ the new slot has been
-		// assigned into the instance, such that its own rendering of children
-		// will be empty (the new Slot will subsume all fills for this name).
-		if ( previousSlot ) {
-			previousSlot.rerender();
-		}
-	}
```

Comment:
All this code collapses into a simple `slots.set` call:
- `triggerListeners` is built-in in `slots.set`.
- the `forceUpdateSlot` logic is built-in into `useSyncExternalStore`. It handles the situation when the store is updated after the consumer component is rendered and before it subscribes to updates in an effect.
- `previousSlot.rerender()` happens automatically because the old slot calls the `currentSlot = useObservableValue( slots, name )` hook and gets notified when a new slot is registered under the same name.

Reply:
Now when we use an observable store for `slots` and observe it with `useSyncExternalStore`, there isn't really much to document. You just use a native hook and it does all the work for you.

The previous implementation was convoluted because:
- the `Slot` component was not subscribed to changes in the `slots` registry. Only `Fill` was. That's why functions in the provider had to manually call `slot.rerender()` on every relevant event. That's not replaced with `useSyncExternalStore`/`useObservableMap` with automatically rerenders the `Slot` component whenever the entry in `slots` changes.
- it manually handled the situation when store was updated between a render (reading a value from store) and the effect that subscribed to updates. Then the update was missed. Before `useSyncExternalStore`, you needed to be aware of such gotchas. For example, the Redux `connect` implementation had to handle this, too. But today we have a native hook that does it all for us. It handles it in this code:

https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberHooks.js#L1900-L1907